ci: Add Windows + UCRT jobs for cross-compiling and native testing

hebasto opened the pull request:
This PR is part of the ongoing effort to migrate to the modern UCRT runtime for cross-compiled Windows binaries, including release builds.

For more details about this migration, see:
- https://github.com/bitcoin/bitcoin/issues/30210
- https://github.com/bitcoin/bitcoin/pull/33593

MSVCRT-related CI jobs should be removed from the CI framework once the migration to UCRT is complete.

maflcko commented on ci/test/00_setup_env_win64_ucrt.sh:
```diff
@@ -0,0 +1,18 @@
+
+export LC_ALL=C.UTF-8
+
+export CONTAINER_NAME=ci_win64_ucrt
```
nit: Could consider naming this `ci_win64` (and the file), as this will be the default eventually, going forward.

hebasto replied:
Thanks! [Reworked](https://github.com/bitcoin/bitcoin/pull/33764#issuecomment-3482770283).

maflcko reviewed (commented):
should the depends build instructions be updated to mention this?

fanquake commented:
> I think it is just

If it had been documented (in the commit or PR description), we wouldn't need to guess. Doing that seems like the minimum that should be done, when adding new suppressions; rather than supplying no information, and claiming someone should look in a followup.

hebasto commented:
> > If this is only appearing for this target, in new code, why not address it here? Would be good to at least document what it is.
> 
> I think it is just one of the many gcc false-positive warnings.

That sounds plausible, though I'm not entirely convinced at this point.

hebasto commented:
From https://github.com/bitcoin/bitcoin/actions/runs/19014374041/job/54300038967:

```
In file included from /usr/lib/gcc/x86_64-w64-mingw32ucrt/14/include/c++/array:43,
                 from /usr/lib/gcc/x86_64-w64-mingw32ucrt/14/include/c++/span:43,
                 from /home/admin/actions-runner/_work/_temp/src/span.h:10,
                 from /home/admin/actions-runner/_work/_temp/src/uint256.h:10,
                 from /home/admin/actions-runner/_work/_temp/src/consensus/params.h:10,
                 from /home/admin/actions-runner/_work/_temp/src/kernel/chainparams.h:9,
                 from /home/admin/actions-runner/_work/_temp/src/chainparams.h:9,
                 from /home/admin/actions-runner/_work/_temp/src/test/net_tests.cpp:5:
In static member function ‘static constexpr _Up* std::__copy_move<_IsMove, true, std::random_access_iterator_tag>::__copy_m(_Tp*, _Tp*, _Up*) [with _Tp = char; _Up = char; bool _IsMove = false]’,
    inlined from ‘constexpr _OI std::__copy_move_a2(_II, _II, _OI) [with bool _IsMove = false; _II = char*; _OI = char*]’ at /usr/lib/gcc/x86_64-w64-mingw32ucrt/14/include/c++/bits/stl_algobase.h:521:30,
    inlined from ‘constexpr _OI std::__copy_move_a1(_II, _II, _OI) [with bool _IsMove = false; _II = char*; _OI = char*]’ at /usr/lib/gcc/x86_64-w64-mingw32ucrt/14/include/c++/bits/stl_algobase.h:548:42,
    inlined from ‘constexpr _OI std::__copy_move_a(_II, _II, _OI) [with bool _IsMove = false; _II = __gnu_cxx::__normal_iterator<char*, __cxx11::basic_string<char> >; _OI = char*]’ at /usr/lib/gcc/x86_64-w64-mingw32ucrt/14/include/c++/bits/stl_algobase.h:555:31,
    inlined from ‘constexpr _OI std::copy(_II, _II, _OI) [with _II = __gnu_cxx::__normal_iterator<char*, __cxx11::basic_string<char> >; _OI = char*]’ at /usr/lib/gcc/x86_64-w64-mingw32ucrt/14/include/c++/bits/stl_algobase.h:651:7,
    inlined from ‘void net_tests::{anonymous}::V2TransportTester::SendMessageA(std::string, std::span<const unsigned char>)’ at /home/admin/actions-runner/_work/_temp/src/test/net_tests.cpp:1304:18:
/usr/lib/gcc/x86_64-w64-mingw32ucrt/14/include/c++/bits/stl_algobase.h:452:30: warning: ‘void* __builtin_memmove(void*, const void*, long long unsigned int)’ offset [0, 1] is out of the bounds [0, 0] [-Warray-bounds=]
  452 |             __builtin_memmove(__result, __first, sizeof(_Tp) * _Num);
      |             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

@sipa

As the [author](https://github.com/bitcoin/bitcoin/commit/91e1ef8684997fb4b3e8b64ef3935a936445066b) of this code, would you consider this warning a false positive?

hebasto commented:
> ... we are considering switching to using LLVM for Windows builds (has that been discussed at length somewhere)?

The only discussion I'm aware of is in https://github.com/bitcoin/bitcoin/issues/31388, and no conclusion has been reached so far.

On the other hand, there seems to be agreement on https://github.com/bitcoin/bitcoin/issues/30210.

hebasto commented:
> I'm not sure whether to Concept ACK/NACK this...

Since this PR is only a subtask of the broader effort to resolve https://github.com/bitcoin/bitcoin/issues/30210, it seems more appropriate to have a conceptual UCRT-related discussion there on in https://github.com/bitcoin/bitcoin/pull/33593.

This one simply addresses your comments [here](https://github.com/bitcoin/bitcoin/pull/33593#issuecomment-3460836590) and [here](https://github.com/bitcoin/bitcoin/pull/33593#pullrequestreview-3398698978).

maflcko commented:
This looks like a nice and small CI-only/test-only change to flatten the way for https://github.com/bitcoin/bitcoin/issues/30210. There is already a pull open for the guix changes in https://github.com/bitcoin/bitcoin/pull/33593, which can then pick up any follow-ups from here and also conclude by removing the msvcrt CI config and msvcrt workarounds in the code.

review ACK 3b135a8fc4451c93b3ea50b3f4621e0d19f35daf 🐂

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 3b135a8fc4451c93b3ea50b3f4621e0d19f35daf 🐂
GGNiE1hD13QtTLvcrvA92ZP1OD3ly24HSGZbRQceL4YLLAM21XtAoV9N81XNmrIwRlFV/bq+Wa7ZbWQPllyIAA==
```

</details>

fanquake commented on .github/workflows/ci.yml:
```diff
@@ -445,6 +437,122 @@ jobs:
           TEST_RUNNER_EXTRA: ${{ github.event_name != 'pull_request' && '--extended' || '' }}
         run: py -3 test/functional/test_runner.py --jobs $NUMBER_OF_PROCESSORS --ci --quiet --tmpdirprefix="$RUNNER_TEMP" --combinedlogslen=99999999 --timeout-factor=$TEST_RUNNER_TIMEOUT_FACTOR $EXCLUDE $TEST_RUNNER_EXTRA
 
+  windows-ucrt-cross:
```
Is there a way to do this without copy-pasting/duplicating > 100 lines of CI config (where both copies need to be kept in sync).

fanquake commented:
> This one simply addresses your comments https://github.com/bitcoin/bitcoin/pull/33593#issuecomment-3460836590 and https://github.com/bitcoin/bitcoin/pull/33593#pullrequestreview-3398698978.

Somewhat, but if this is the approach, it means the not only the compiler, but also the version of the headers being tested isn't going to match Guix? (I'm wondering if internalising the headers might be a better approach, as that would be more flexible, and, if there is some plan to migrate to Clang, would likely be needed in any case?).

maflcko commented:
I think we've never seen a  header-version specific bug in the windows-cross compilation, so trying to fit the major versions in the CI on a best-effort basis seems sufficient for now? If more specific matching is needed for better testing a specific scenario, or for llvm builds, it seems sufficient to do as a follow-up?

hebasto commented:
> > should the depends build instructions be updated to mention this?
> 
> I'd prefer updating the docs once `depends/hosts/mingw32.mk` is adjusted so that compilers don't need to be specified explicitly.

Done in https://github.com/bitcoin/bitcoin/pull/33857.

hebasto commented:
Rebased.

maflcko reviewed (approved):
lgtm

hebasto committed:
ci: Rename items specific to Windows + MSVCRT

This is necessary to prepare for introducing the new Windows + UCRT
script and jobs.

hebasto commented on ci/test/00_setup_env_win64_msvcrt.sh:
```diff
@@ -0,0 +1,22 @@
+
+export LC_ALL=C.UTF-8
+
+export CONTAINER_NAME=ci_win64
```
Thanks! Fixed.

fanquake reviewed (changes requested):
https://github.com/bitcoin/bitcoin/pull/33764#discussion_r2549961718

hebasto commented on ci/test/00_setup_env_win64.sh:
```diff
@@ -1,22 +1,18 @@
 export RUN_FUNCTIONAL_TESTS=false
 export GOAL="deploy"
-export BITCOIN_CONFIG="\
-  --preset=dev-mode \
```
> Yeah, I guess there were some silent conflicts.

Indeed. Fixed.

hebasto commented:
> [#33764 (comment)](https://github.com/bitcoin/bitcoin/pull/33764#discussion_r2549961718)

Thanks! Fixed.

hebasto commented:
All conflicts caused by the recent rebase have been fixed, and CI is green now.

hebasto committed:
ci: Add Windows + UCRT jobs for cross-compiling and native testing

Co-authored-by: will <will@256k1.dev>

hebasto commented:
@willcl-ark's [suggestion](https://github.com/bitcoin/bitcoin/pull/33764#discussion_r2556213666) has been taken.

Thank you!

maflcko commented:
review ACK 2e27bd9c3af91eb9fcc626fe65d065df0a80974d 🖊

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 2e27bd9c3af91eb9fcc626fe65d065df0a80974d 🖊
VdW8j9UHIembvFBZW5872vAJlaVGWUbBOGG1CuEaRmfMCYMPTi086fRbdP5DzDkwdOOSZVikVeX0e86X9RQWCQ==
```

</details>
